File: 29Items-Projects/7Algorithms/6-Backtracking/python/sudoku_solver.py
"""Sudoku Solver — classic 9×9 backtracking solver."""


class SudokuSolver:
    def __init__(self, board: list[list[int]]):
        self.board = [row[:] for row in board]

    def solve_and_print(self) -> None:
        print("Input puzzle:")
        self._print_board()

        solved = self._solve(0, 0)
        print("\nSolved:" if solved else "\nNo solution exists.")
        if solved:
            self._print_board()

    def _solve(self, row: int, col: int) -> bool:
        if row == 9:
            return True
        next_row = row + 1 if col == 8 else row
        next_col = (col + 1) % 9

        if self.board[row][col] != 0:
            return self._solve(next_row, next_col)

        for num in range(1, 10):
            if self._is_valid(row, col, num):
                self.board[row][col] = num
                if self._solve(next_row, next_col):
                    return True
                self.board[row][col] = 0
        return False

    def _is_valid(self, row: int, col: int, num: int) -> bool:
        for i in range(9):
            if self.board[row][i] == num or self.board[i][col] == num:
                return False
        br, bc = (row // 3) * 3, (col // 3) * 3
        for r in range(br, br + 3):
            for c in range(bc, bc + 3):
                if self.board[r][c] == num:
                    return False
        return True

    def _print_board(self) -> None:
        print("┌───────┬───────┬───────┐")
        for r in range(9):
            if r in (3, 6):
                print("├───────┼───────┼───────┤")
            row = "│"
            for c in range(9):
                cell = "." if self.board[r][c] == 0 else str(self.board[r][c])
                row += f" {cell} "
                if c in (2, 5):
                    row += "│"
            row += "│"
            print(row)
        print("└───────┴───────┴───────┘")

File: 29Items-Projects/7Algorithms/6-Backtracking/java/src/sudoku/SudokuSolver.java
package sudoku;

public class SudokuSolver {

    private final int[][] board;
    private boolean solved = false;

    public SudokuSolver(int[][] board) {
        this.board = new int[9][9];
        for (int i = 0; i < 9; i++) {
            this.board[i] = board[i].clone();
        }
    }

    public void solveAndPrint() {
        System.out.println("Input puzzle:");
        printBoard();

        solved = solve(0, 0);

        if (solved) {
            System.out.println("\nSolved:");
            printBoard();
        } else {
            System.out.println("\nNo solution exists.");
        }
    }

    private boolean solve(int row, int col) {
        if (row == 9) return true;
        int nextRow = (col == 8) ? row + 1 : row;
        int nextCol = (col + 1) % 9;

        if (board[row][col] != 0) {
            return solve(nextRow, nextCol);
        }

        for (int num = 1; num <= 9; num++) {
            if (isValid(row, col, num)) {
                board[row][col] = num;
                if (solve(nextRow, nextCol)) return true;
                board[row][col] = 0;
            }
        }
        return false;
    }

    private boolean isValid(int row, int col, int num) {
        for (int i = 0; i < 9; i++) {
            if (board[row][i] == num || board[i][col] == num) return false;
        }
        int boxRow = (row / 3) * 3, boxCol = (col / 3) * 3;
        for (int r = boxRow; r < boxRow + 3; r++) {
            for (int c = boxCol; c < boxCol + 3; c++) {
                if (board[r][c] == num) return false;
            }
        }
        return true;
    }

    private void printBoard() {
        System.out.println("+-------+-------+-------+");
        for (int r = 0; r < 9; r++) {
            if (r == 3 || r == 6) System.out.println("+-------+-------+-------+");
            StringBuilder sb = new StringBuilder("|");
            for (int c = 0; c < 9; c++) {
                sb.append(board[r][c] == 0 ? " . " : " " + board[r][c] + " ");
                if (c == 2 || c == 5) sb.append("|");
            }
            sb.append("|");
            System.out.println(sb);
        }
        System.out.println("+-------+-------+-------+");
    }
}
